inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java

```diff
@@ -430,13 +431,29 @@ public boolean existsRecommender(Project aProject, String aName)
     public void onDocumentOpened(DocumentOpenedEvent aEvent)
     {
         Project project = aEvent.getDocument().getProject();
-        String user = aEvent.getAnnotator();
+        String username = aEvent.getAnnotator();
+        SourceDocument doc = aEvent.getDocument();
 
         // If there already is a state, we just re-use it. We only trigger a new training if there
         // is no state yet.
-        if (!states.containsKey(new RecommendationStateKey(user, project))) {
-            triggerTrainingAndClassification(user, project, "DocumentOpenedEvent");
+        if (!states.containsKey(new RecommendationStateKey(username, project))) {
+            triggerTrainingAndClassification(username, project, "DocumentOpenedEvent", doc);
+        } else {
+            // If we already trained, predicted only for the last document and open a new
+            // document, we start the predictions so that the user gets recommendations
+            // as quickly as possible without any interaction needed
+            User user = userRepository.get(username);
+            Predictions predictions = getPredictions(user, project);
+            if (predictions == null || !predictions.hasPredictions()) {
+                log.debug("Starting prediction task after document was opened!");
+                Task task = new PredictionTask(user, project, "DocumentOpenedEvent", doc);
+                schedulingService.enqueue(task);
+            }
         }
+
+        // We reset this in case the state was not properly cleared, e.g. the AL session
```

Comment:
I think it might work, but it would be good to have extensive explanatory comments in both places in the code including a rationale why it should work and a cross-reference.

Reply:
@jcklie before merging this, could you expand the comments in these locations (see my comment above)?

Reply:
I added some minor additions and cross referenced them.